File: mini-agent-harness/python/mcp_session.py
from __future__ import annotations

from copy import deepcopy
from dataclasses import dataclass
from types import MappingProxyType
from typing import Awaitable, Callable, Literal, Mapping, Protocol
import json
import re


class CancellationView(Protocol):
    @property
    def cancelled(self) -> bool: ...
    def throw_if_cancelled(self) -> None: ...


@dataclass(frozen=True)
class McpServerCapabilities:
    tools: bool
    resources: bool
    prompts: bool


@dataclass(frozen=True)
class McpHandshake:
    server_name: str
    server_version: str
    capabilities: McpServerCapabilities
    instructions: str | None = None


@dataclass(frozen=True)
class McpRemoteTool:
    name: str
    description: str
    input_schema: Mapping[str, object]
    annotations: Mapping[str, object] | None = None


@dataclass(frozen=True)
class McpCallResult:
    content: object
    is_error: bool = False


class McpTransport(Protocol):
    async def connect(self, signal: CancellationView) -> McpHandshake: ...
    async def list_tools(
        self, signal: CancellationView
    ) -> tuple[McpRemoteTool, ...]: ...
    async def call_tool(
        self,
        tool_name: str,
        arguments: Mapping[str, object],
        *,
        signal: CancellationView,
        idempotency_key: str,
        timeout_ms: int,
    ) -> McpCallResult: ...
    async def close(self, reason: str) -> None: ...


@dataclass(frozen=True)
class McpVisibleTool:
    qualified_name: str
    remote_name: str
    description: str
    input_schema: Mapping[str, object]


@dataclass(frozen=True)
class McpCapabilitySnapshot:
    server_name: str
    generation: int
    revision: int
    capabilities: McpServerCapabilities
    tools: tuple[McpVisibleTool, ...]


@dataclass(frozen=True)
class McpSessionState:
    type: Literal["idle", "connecting", "ready", "degraded", "closed"]
    generation: int
    revision: int = 0
    reason: str | None = None


@dataclass(frozen=True)
class McpSessionTrace:
    action: Literal["connect", "refresh", "degrade", "disconnect", "call", "retry"]
    generation: int
    revision: int
    tool_count: int


class McpRetryPolicy(Protocol):
    def allow_session_recovery(self, tool: McpRemoteTool) -> bool: ...


class NoMcpRetryPolicy:
    def allow_session_recovery(self, _tool: McpRemoteTool) -> bool:
        return False


class McpSessionExpiredError(RuntimeError):
    pass


class StaleMcpSnapshotError(RuntimeError):
    pass


class IndeterminateMcpOutcomeError(RuntimeError):
    pass


class McpSession:
    def __init__(
        self,
        factory: Callable[[], McpTransport],
        *,
        retry_policy: McpRetryPolicy | None = None,
        timeout_ms: int = 60_000,
    ) -> None:
        if timeout_ms <= 0:
            raise ValueError("MCP timeout must be positive")
        self._factory = factory
        self._retry_policy = retry_policy or NoMcpRetryPolicy()
        self._timeout_ms = timeout_ms
        self._state = McpSessionState("idle", 0)
        self._revision = 0
        self._transport: McpTransport | None = None
        self._handshake: McpHandshake | None = None
        self._tools: tuple[McpRemoteTool, ...] = ()
        self._traces: list[McpSessionTrace] = []

    @property
    def state(self) -> McpSessionState:
        return self._state

    async def connect(self, signal: CancellationView) -> McpCapabilitySnapshot:
        signal.throw_if_cancelled()
        if self._state.type == "closed":
            raise RuntimeError("MCP session is closed")
        generation = self._state.generation + 1
        self._state = McpSessionState("connecting", generation, self._revision)
        candidate = self._factory()
        try:
            handshake = _validate_handshake(await candidate.connect(signal))
            signal.throw_if_cancelled()
            tools = (
                _validate_tools(await candidate.list_tools(signal))
                if handshake.capabilities.tools
                else ()
            )
            signal.throw_if_cancelled()
            if self._transport is not None:
                await self._transport.close("replaced by newer MCP generation")
            self._transport = candidate
            self._handshake = handshake
            self._tools = tools
            self._revision += 1
            self._state = McpSessionState("ready", generation, self._revision)
            self._record("connect")
            return self.snapshot()
        except Exception as error:
            try:
                await candidate.close("MCP initialization failed")
            except Exception:
                pass
            self._state = McpSessionState(
                "degraded", generation, self._revision, type(error).__name__
            )
            self._record("degrade")
            raise

    def snapshot(self) -> McpCapabilitySnapshot:
        if self._state.type != "ready" or self._handshake is None:
            raise RuntimeError("MCP session is not ready")
        return McpCapabilitySnapshot(
            self._handshake.server_name,
            self._state.generation,
            self._revision,
            self._handshake.capabilities,
            tuple(
                McpVisibleTool(
                    _qualify(self._handshake.server_name, tool.name),
                    tool.name,
                    tool.description,
                    _freeze_mapping(tool.input_schema),
                )
                for tool in self._tools
            ),
        )

    async def handle_tools_changed(
        self, signal: CancellationView
    ) -> McpCapabilitySnapshot:
        if self._state.type != "ready" or self._transport is None:
            raise RuntimeError("MCP session is not ready")
        try:
            tools = _validate_tools(await self._transport.list_tools(signal))
            signal.throw_if_cancelled()
            self._tools = tools
            self._revision += 1
            self._state = McpSessionState(
                "ready", self._state.generation, self._revision
            )
            self._record("refresh")
            return self.snapshot()
        except Exception as error:
            self._state = McpSessionState(
                "degraded",
                self._state.generation,
                self._revision,
                type(error).__name__,
            )
            self._record("degrade")
            raise

    async def call(
        self,
        snapshot: McpCapabilitySnapshot,
        qualified_name: str,
        arguments: Mapping[str, object],
        call_id: str,
        signal: CancellationView,
    ) -> McpCallResult:
        tool = self._assert_callable(snapshot, qualified_name)
        idempotency_key = (
            f"{snapshot.server_name}:{snapshot.generation}:{call_id}"
        )
        try:
            assert self._transport is not None
            result = await self._transport.call_tool(
                tool.name,
                _freeze_mapping(arguments),
                signal=signal,
                idempotency_key=idempotency_key,
                timeout_ms=self._timeout_ms,
            )
            self._record("call")
            return result
        except McpSessionExpiredError as error:
            if not self._retry_policy.allow_session_recovery(tool):
                raise IndeterminateMcpOutcomeError(
                    f"MCP call outcome is indeterminate and retry is not approved: {qualified_name}"
                ) from error
            recovered = await self.connect(signal)
            recovered_tool = next(
                (
                    item
                    for item in self._tools
                    if _qualify(recovered.server_name, item.name) == qualified_name
                ),
                None,
            )
            if recovered_tool is None or _stable_schema(
                recovered_tool.input_schema
            ) != _stable_schema(tool.input_schema):
                raise IndeterminateMcpOutcomeError(
                    "MCP tool changed during session recovery"
                )
            self._record("retry")
            assert self._transport is not None
            return await self._transport.call_tool(
                recovered_tool.name,
                _freeze_mapping(arguments),
                signal=signal,
                idempotency_key=idempotency_key,
                timeout_ms=self._timeout_ms,
            )

    async def disconnect(self, reason: str = "MCP disconnect") -> None:
        if self._transport is not None:
            await self._transport.close(reason)
        self._transport = None
        self._handshake = None
        self._tools = ()
        self._state = McpSessionState("idle", self._state.generation, self._revision)
        self._record("disconnect")

    def traces(self) -> tuple[McpSessionTrace, ...]:
        return tuple(self._traces)

    def _assert_callable(
        self, snapshot: McpCapabilitySnapshot, qualified_name: str
    ) -> McpRemoteTool:
        if (
            self._state.type != "ready"
            or self._transport is None
            or self._handshake is None
        ):
            raise RuntimeError("MCP session is not ready")
        if (
            snapshot.generation != self._state.generation
            or snapshot.revision != self._revision
        ):
            raise StaleMcpSnapshotError("MCP capability snapshot is stale")
        descriptor = next(
            (item for item in snapshot.tools if item.qualified_name == qualified_name),
            None,
        )
        if descriptor is None:
            raise ValueError(f"MCP tool is not visible: {qualified_name}")
        tool = next(
            (item for item in self._tools if item.name == descriptor.remote_name),
            None,
        )
        if tool is None:
            raise ValueError(f"MCP tool has no remote handler: {qualified_name}")
        return tool

    def _record(self, action) -> None:
        self._traces.append(
            McpSessionTrace(
                action,
                self._state.generation,
                self._revision,
                len(self._tools),
            )
        )


def _validate_handshake(value: McpHandshake) -> McpHandshake:
    _require_text(value.server_name, "MCP server name")
    _require_text(value.server_version, "MCP server version")
    return value


def _validate_tools(
    tools: tuple[McpRemoteTool, ...]
) -> tuple[McpRemoteTool, ...]:
    names: set[str] = set()
    frozen: list[McpRemoteTool] = []
    for tool in tools:
        _require_text(tool.name, "MCP tool name")
        _require_text(tool.description, "MCP tool description")
        if tool.name in names:
            raise ValueError(f"duplicate MCP tool: {tool.name}")
        names.add(tool.name)
        if tool.input_schema.get("type") not in (None, "object"):
            raise ValueError(
                f"MCP tool schema must describe an object: {tool.name}"
            )
        frozen.append(
            McpRemoteTool(
                tool.name,
                tool.description,
                _freeze_mapping(tool.input_schema),
                _freeze_mapping(tool.annotations) if tool.annotations else None,
            )
        )
    return tuple(frozen)


def _qualify(server_name: str, tool_name: str) -> str:
    return f"mcp__{_normalize(server_name)}__{_normalize(tool_name)}"


def _normalize(value: str) -> str:
    return re.sub(r"[^A-Za-z0-9_-]", "_", value)


def _stable_schema(value: Mapping[str, object]) -> str:
    return json.dumps(dict(value), sort_keys=True, separators=(",", ":"))


def _freeze_mapping(value: Mapping[str, object]) -> Mapping[str, object]:
    return MappingProxyType(deepcopy(dict(value)))


def _require_text(value: str, label: str) -> None:
    if not value.strip():
        raise ValueError(f"{label} must not be empty")

File: mini-agent-harness/typescript/agent/mcpSession.ts
export type McpServerCapabilities = Readonly<{
  tools: boolean
  resources: boolean
  prompts: boolean
}>

export type McpHandshake = Readonly<{
  serverName: string
  serverVersion: string
  capabilities: McpServerCapabilities
  instructions?: string
}>

export type McpRemoteTool = Readonly<{
  name: string
  description: string
  inputSchema: Readonly<Record<string, unknown>>
  annotations?: Readonly<Record<string, unknown>>
}>

export type McpCallResult = Readonly<{
  content: unknown
  isError?: boolean
}>

export type McpTransportNotification = 'tools/list_changed'

export interface McpTransport {
  connect(signal: AbortSignal): Promise<McpHandshake>
  listTools(signal: AbortSignal): Promise<readonly McpRemoteTool[]>
  callTool(
    toolName: string,
    args: Readonly<Record<string, unknown>>,
    options: Readonly<{ signal: AbortSignal; idempotencyKey: string; timeoutMs: number }>,
  ): Promise<McpCallResult>
  close(reason: string): Promise<void>
}

export type McpTransportFactory = () => McpTransport

export type McpSessionState =
  | Readonly<{ type: 'idle'; generation: number }>
  | Readonly<{ type: 'connecting'; generation: number }>
  | Readonly<{ type: 'ready'; generation: number; revision: number }>
  | Readonly<{ type: 'degraded'; generation: number; revision: number; reason: string }>
  | Readonly<{ type: 'closed'; generation: number }>

export type McpCapabilitySnapshot = Readonly<{
  serverName: string
  generation: number
  revision: number
  capabilities: McpServerCapabilities
  tools: readonly Readonly<{
    qualifiedName: string
    remoteName: string
    description: string
    inputSchema: Readonly<Record<string, unknown>>
  }>[]
}>

export type McpSessionTrace = Readonly<{
  action: 'connect' | 'refresh' | 'degrade' | 'disconnect' | 'call' | 'retry'
  generation: number
  revision: number
  toolCount: number
}>

export interface McpRetryPolicy {
  allowSessionRecovery(tool: McpRemoteTool): boolean
}

export class NoMcpRetryPolicy implements McpRetryPolicy {
  allowSessionRecovery(): boolean { return false }
}

export class McpSessionExpiredError extends Error {}
export class StaleMcpSnapshotError extends Error {}
export class IndeterminateMcpOutcomeError extends Error {}

export class McpSession {
  #state: McpSessionState = Object.freeze({ type: 'idle', generation: 0 })
  #revision = 0
  #transport?: McpTransport
  #handshake?: McpHandshake
  #tools: readonly McpRemoteTool[] = Object.freeze([])
  readonly #factory: McpTransportFactory
  readonly #retryPolicy: McpRetryPolicy
  readonly #timeoutMs: number
  readonly #trace: McpSessionTrace[] = []

  constructor(
    factory: McpTransportFactory,
    options: Readonly<{ retryPolicy?: McpRetryPolicy; timeoutMs?: number }> = {},
  ) {
    this.#factory = factory
    this.#retryPolicy = options.retryPolicy ?? new NoMcpRetryPolicy()
    this.#timeoutMs = options.timeoutMs ?? 60_000
    if (!Number.isInteger(this.#timeoutMs) || this.#timeoutMs <= 0) {
      throw new Error('MCP timeout must be a positive integer')
    }
  }

  state(): McpSessionState { return deepFreeze(structuredClone(this.#state)) }

  async connect(signal: AbortSignal): Promise<McpCapabilitySnapshot> {
    throwIfAborted(signal)
    if (this.#state.type === 'closed') throw new Error('MCP session is closed')
    const generation = this.#state.generation + 1
    this.#state = Object.freeze({ type: 'connecting', generation })
    const candidate = this.#factory()
    try {
      const handshake = validateHandshake(await candidate.connect(signal))
      throwIfAborted(signal)
      const tools = handshake.capabilities.tools
        ? validateTools(await candidate.listTools(signal))
        : Object.freeze([])
      throwIfAborted(signal)
      await this.#transport?.close('replaced by newer MCP generation')
      this.#transport = candidate
      this.#handshake = handshake
      this.#tools = tools
      this.#revision += 1
      this.#state = Object.freeze({ type: 'ready', generation, revision: this.#revision })
      this.#record('connect')
      return this.snapshot()
    } catch (error) {
      await candidate.close('MCP initialization failed').catch(() => {})
      this.#state = Object.freeze({
        type: 'degraded', generation, revision: this.#revision, reason: safeCategory(error),
      })
      this.#record('degrade')
      throw error
    }
  }

  snapshot(): McpCapabilitySnapshot {
    if (this.#state.type !== 'ready' || !this.#handshake) {
      throw new Error('MCP session is not ready')
    }
    return deepFreeze({
      serverName: this.#handshake.serverName,
      generation: this.#state.generation,
      revision: this.#revision,
      capabilities: this.#handshake.capabilities,
      tools: this.#tools.map(tool => ({
        qualifiedName: qualify(this.#handshake!.serverName, tool.name),
        remoteName: tool.name,
        description: tool.description,
        inputSchema: tool.inputSchema,
      })),
    })
  }

  async handleNotification(
    notification: McpTransportNotification,
    signal: AbortSignal,
  ): Promise<McpCapabilitySnapshot> {
    if (notification !== 'tools/list_changed') throw new Error('unsupported MCP notification')
    if (this.#state.type !== 'ready' || !this.#transport) {
      throw new Error('MCP session is not ready')
    }
    try {
      const tools = validateTools(await this.#transport.listTools(signal))
      throwIfAborted(signal)
      this.#tools = tools
      this.#revision += 1
      this.#state = Object.freeze({
        type: 'ready', generation: this.#state.generation, revision: this.#revision,
      })
      this.#record('refresh')
      return this.snapshot()
    } catch (error) {
      this.#state = Object.freeze({
        type: 'degraded', generation: this.#state.generation,
        revision: this.#revision, reason: safeCategory(error),
      })
      this.#record('degrade')
      throw error
    }
  }

  async call(
    snapshot: McpCapabilitySnapshot,
    qualifiedName: string,
    args: Readonly<Record<string, unknown>>,
    callId: string,
    signal: AbortSignal,
  ): Promise<McpCallResult> {
    const tool = this.#assertCallable(snapshot, qualifiedName)
    const idempotencyKey = `${snapshot.serverName}:${snapshot.generation}:${callId}`
    try {
      const result = await this.#transport!.callTool(tool.name, structuredClone(args), {
        signal, idempotencyKey, timeoutMs: this.#timeoutMs,
      })
      this.#record('call')
      return deepFreeze(structuredClone(result))
    } catch (error) {
      if (!(error instanceof McpSessionExpiredError)) throw error
      if (!this.#retryPolicy.allowSessionRecovery(tool)) {
        throw new IndeterminateMcpOutcomeError(
          `MCP call outcome is indeterminate and retry is not approved: ${qualifiedName}`,
        )
      }
      const recovered = await this.connect(signal)
      const recoveredTool = this.#tools.find(item =>
        qualify(recovered.serverName, item.name) === qualifiedName,
      )
      if (!recoveredTool || stableSchema(recoveredTool.inputSchema) !== stableSchema(tool.inputSchema)) {
        throw new IndeterminateMcpOutcomeError('MCP tool changed during session recovery')
      }
      this.#record('retry')
      return deepFreeze(structuredClone(await this.#transport!.callTool(
        recoveredTool.name, structuredClone(args),
        { signal, idempotencyKey, timeoutMs: this.#timeoutMs },
      )))
    }
  }

  async disconnect(reason = 'MCP disconnect'): Promise<void> {
    await this.#transport?.close(reason)
    this.#transport = undefined
    this.#handshake = undefined
    this.#tools = Object.freeze([])
    this.#state = Object.freeze({ type: 'idle', generation: this.#state.generation })
    this.#record('disconnect')
  }

  async close(): Promise<void> {
    await this.#transport?.close('MCP session closed')
    this.#transport = undefined
    this.#handshake = undefined
    this.#tools = Object.freeze([])
    this.#state = Object.freeze({ type: 'closed', generation: this.#state.generation })
  }

  traces(): readonly McpSessionTrace[] { return deepFreeze(this.#trace.map(item => ({ ...item }))) }

  #assertCallable(snapshot: McpCapabilitySnapshot, qualifiedName: string): McpRemoteTool {
    if (this.#state.type !== 'ready' || !this.#transport || !this.#handshake) {
      throw new Error('MCP session is not ready')
    }
    if (snapshot.generation !== this.#state.generation || snapshot.revision !== this.#revision) {
      throw new StaleMcpSnapshotError('MCP capability snapshot is stale')
    }
    const descriptor = snapshot.tools.find(item => item.qualifiedName === qualifiedName)
    if (!descriptor) throw new Error(`MCP tool is not visible: ${qualifiedName}`)
    const tool = this.#tools.find(item => item.name === descriptor.remoteName)
    if (!tool) throw new Error(`MCP tool has no remote handler: ${qualifiedName}`)
    return tool
  }

  #record(action: McpSessionTrace['action']): void {
    this.#trace.push(Object.freeze({
      action, generation: this.#state.generation, revision: this.#revision,
      toolCount: this.#tools.length,
    }))
  }
}

function validateHandshake(value: McpHandshake): McpHandshake {
  requireText(value.serverName, 'MCP server name')
  requireText(value.serverVersion, 'MCP server version')
  return deepFreeze(structuredClone(value))
}

function validateTools(tools: readonly McpRemoteTool[]): readonly McpRemoteTool[] {
  const names = new Set<string>()
  return deepFreeze(tools.map(tool => {
    requireText(tool.name, 'MCP tool name')
    requireText(tool.description, 'MCP tool description')
    if (names.has(tool.name)) throw new Error(`duplicate MCP tool: ${tool.name}`)
    names.add(tool.name)
    if (tool.inputSchema.type !== undefined && tool.inputSchema.type !== 'object') {
      throw new Error(`MCP tool schema must describe an object: ${tool.name}`)
    }
    return structuredClone(tool)
  }))
}

function qualify(serverName: string, toolName: string): string {
  return `mcp__${normalize(serverName)}__${normalize(toolName)}`
}

function normalize(value: string): string {
  return value.replace(/[^A-Za-z0-9_-]/g, '_')
}

function stableSchema(value: Readonly<Record<string, unknown>>): string {
  return JSON.stringify(value, Object.keys(value).sort())
}

function throwIfAborted(signal: AbortSignal): void {
  if (signal.aborted) throw signal.reason ?? new DOMException('Aborted', 'AbortError')
}

function safeCategory(error: unknown): string {
  return error instanceof Error ? error.name : 'UnknownError'
}

function requireText(value: string, label: string): void {
  if (!value.trim()) throw new Error(`${label} must not be empty`)
}

function deepFreeze<T>(value: T): T {
  if (value !== null && typeof value === 'object' && !Object.isFrozen(value)) {
    for (const nested of Object.values(value as Record<string, unknown>)) deepFreeze(nested)
    Object.freeze(value)
  }
  return value
}
